edgedb-protocol/src/model/time.rs

```diff
@@ -47,26 +80,339 @@ impl Duration {
 }
 
 impl LocalDatetime {
+    pub const MIN : LocalDatetime = LocalDatetime { micros: LocalDate::MIN.days as i64 * MICROS_PER_DAY as i64 };
+    pub const MAX : LocalDatetime = LocalDatetime { 
+        micros: LocalDate::MAX.days as i64 * MICROS_PER_DAY as i64
+         + LocalTime::MAX.micros as i64
+    };
+
+    fn try_from_micros(micros: i64) -> Result<LocalDatetime, OutOfRangeError> {
+        if micros < Self::MIN.micros || micros > Self::MAX.micros {
+            return Err(OutOfRangeError);
+        }
+        Ok(LocalDatetime { micros })
+    }
+
     pub fn from_micros(micros: i64) -> LocalDatetime {
-        return LocalDatetime { micros }
+        Self::try_from_micros(micros).expect(&format!(
+            "LocalDatetime::from_micros({}) is outside the valid datetime range",
+             micros))
+    }
+
+    pub fn to_micros(self) -> i64 {
+        self.micros
+    }
+
+    pub fn new(date: LocalDate, time: LocalTime) -> LocalDatetime {
+        Self::from_micros(date.to_days() as i64 * MICROS_PER_DAY as i64 + time.to_micros() as i64)
+    }
+
+    pub fn date(self) -> LocalDate {
+        LocalDate::from_days(self.micros.wrapping_div_euclid(MICROS_PER_DAY as i64) as i32)
```

Comment:
Is this `wrapping_` because it's supposed to be faster?

Reply:
It's `div_euclid` for correctness on negative numbers. I believe the different variants of `div_euclid` only differ if the right side is `-1` which is never the case, so the choice doesn't matter.